File: mythos/celtic/audit_links.py
#!/usr/bin/env python3
"""
Celtic Mythology Link Checker
Finds all broken internal links in the Celtic mythology section
"""

import os
import re
from pathlib import Path
from collections import defaultdict

def find_html_files(directory):
    """Find all HTML files in the directory"""
    return list(Path(directory).rglob("*.html"))

def extract_links(html_content, file_path):
    """Extract all internal links from HTML content"""
    # Match href attributes
    link_pattern = r'href=["\']((?!http|#)[^"\']+)["\']'
    links = re.findall(link_pattern, html_content)

    # Filter out fragments (links starting with #)
    internal_links = [link for link in links if not link.startswith('#')]

    return internal_links

def resolve_link(source_file, link):
    """Resolve a relative link to an absolute path"""
    source_dir = os.path.dirname(source_file)

    # Handle relative paths
    resolved = os.path.normpath(os.path.join(source_dir, link))
    return resolved

def check_link_exists(target_path):
    """Check if the target file exists"""
    return os.path.exists(target_path)

def audit_celtic_links(celtic_dir):
    """Audit all links in Celtic mythology pages"""
    print("=" * 80)
    print("CELTIC MYTHOLOGY - INTERNAL LINK AUDIT")
    print("=" * 80)
    print()

    html_files = find_html_files(celtic_dir)
    print(f"Found {len(html_files)} HTML files in Celtic section\n")

    broken_links = []
    total_links = 0
    files_checked = 0

    for html_file in sorted(html_files):
        files_checked += 1
        try:
            with open(html_file, 'r', encoding='utf-8') as f:
                content = f.read()
        except Exception as e:
            print(f"ERROR reading {html_file}: {e}")
            continue

        links = extract_links(content, str(html_file))

        if links:
            file_has_broken = False
            for link in links:
                total_links += 1
                target = resolve_link(str(html_file), link)

                if not check_link_exists(target):
                    if not file_has_broken:
                        print(f"\n{html_file.relative_to(celtic_dir)}:")
                        file_has_broken = True

                    print(f"  ❌ BROKEN: {link}")
                    print(f"     → Resolves to: {target}")
                    broken_links.append({
                        'source': str(html_file),
                        'link': link,
                        'target': target
                    })

    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print(f"Files checked: {files_checked}")
    print(f"Total internal links: {total_links}")
    print(f"Broken links: {len(broken_links)}")
    print(f"Success rate: {((total_links - len(broken_links)) / total_links * 100):.1f}%")

    if broken_links:
        print("\n⚠️  ACTION REQUIRED: Fix broken links listed above")
    else:
        print("\n✅ All internal links are valid!")

    return broken_links

if __name__ == "__main__":
    celtic_dir = Path(__file__).parent
    audit_celtic_links(celtic_dir)

File: mythos/greek/audit-links.js
// Greek Mythology Section - Link Audit Script
// Checks all internal links within mythos/greek/ directory

const fs = require('fs');
const path = require('path');

const baseDir = path.join(__dirname);
const results = {
    brokenLinks: [],
    workingLinks: [],
    externalLinks: [],
    totalFiles: 0,
    totalLinks: 0
};

// Get all HTML files recursively
function getHtmlFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);

    files.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);

        if (stat.isDirectory()) {
            getHtmlFiles(filePath, fileList);
        } else if (file.endsWith('.html')) {
            fileList.push(filePath);
        }
    });

    return fileList;
}

// Extract all links from HTML content
function extractLinks(html, sourceFile) {
    const links = [];

    // Match href attributes
    const hrefRegex = /href=["']([^"']+)["']/g;
    let match;

    while ((match = hrefRegex.exec(html)) !== null) {
        const href = match[1];

        // Skip anchors, javascript, mailto, etc.
        if (href.startsWith('#') ||
            href.startsWith('javascript:') ||
            href.startsWith('mailto:') ||
            href.startsWith('tel:')) {
            continue;
        }

        links.push({
            href,
            source: sourceFile,
            isExternal: href.startsWith('http://') || href.startsWith('https://') || href.startsWith('//')
        });
    }

    return links;
}

// Check if a relative link target exists
function checkLink(link, sourceFile) {
    if (link.isExternal) {
        results.externalLinks.push({
            source: path.relative(baseDir, sourceFile),
            target: link.href
        });
        return true;
    }

    // Resolve relative path
    const sourceDir = path.dirname(sourceFile);
    let targetPath = link.href;

    // Remove query strings and hash fragments
    targetPath = targetPath.split('?')[0].split('#')[0];

    // Skip empty links
    if (!targetPath || targetPath === '') {
        return true;
    }

    // Handle absolute paths from site root
    if (targetPath.startsWith('../../')) {
        // These are links outside greek section - we'll validate they exist
        const fullPath = path.resolve(sourceDir, targetPath);
        const exists = fs.existsSync(fullPath);

        if (exists) {
            results.workingLinks.push({
                source: path.relative(baseDir, sourceFile),
                target: link.href,
                type: 'external-to-greek'
            });
        } else {
            results.brokenLinks.push({
                source: path.relative(baseDir, sourceFile),
                target: link.href,
                resolvedPath: fullPath,
                type: 'external-to-greek'
            });
        }

        return exists;
    }

    // Resolve the target file path
    const fullTargetPath = path.resolve(sourceDir, targetPath);
    const exists = fs.existsSync(fullTargetPath);

    if (exists) {
        results.workingLinks.push({
            source: path.relative(baseDir, sourceFile),
            target: link.href,
            type: 'internal'
        });
    } else {
        results.brokenLinks.push({
            source: path.relative(baseDir, sourceFile),
            target: link.href,
            resolvedPath: fullTargetPath,
            type: 'internal'
        });
    }

    return exists;
}

// Main audit function
function auditLinks() {
    console.log('Starting Greek Mythology Link Audit...\n');

    const htmlFiles = getHtmlFiles(baseDir);
    results.totalFiles = htmlFiles.length;

    console.log(`Found ${htmlFiles.length} HTML files\n`);

    htmlFiles.forEach(file => {
        const html = fs.readFileSync(file, 'utf8');
        const links = extractLinks(html, file);

        results.totalLinks += links.length;

        links.forEach(link => {
            checkLink(link, file);
        });
    });

    // Print results
    console.log('='.repeat(80));
    console.log('LINK AUDIT RESULTS');
    console.log('='.repeat(80));
    console.log(`\nTotal HTML files scanned: ${results.totalFiles}`);
    console.log(`Total links found: ${results.totalLinks}`);
    console.log(`Working internal links: ${results.workingLinks.filter(l => l.type === 'internal').length}`);
    console.log(`Working external-to-greek links: ${results.workingLinks.filter(l => l.type === 'external-to-greek').length}`);
    console.log(`External links (not checked): ${results.externalLinks.length}`);
    console.log(`\n🔴 BROKEN LINKS: ${results.brokenLinks.length}\n`);

    if (results.brokenLinks.length > 0) {
        console.log('BROKEN LINKS DETAILS:');
        console.log('-'.repeat(80));

        // Group by source file
        const bySource = {};
        results.brokenLinks.forEach(link => {
            if (!bySource[link.source]) {
                bySource[link.source] = [];
            }
            bySource[link.source].push(link);
        });

        Object.keys(bySource).sort().forEach(source => {
            console.log(`\n📄 ${source}`);
            bySource[source].forEach(link => {
                console.log(`   ❌ ${link.target}`);
                console.log(`      → Resolved to: ${link.resolvedPath}`);
            });
        });
    } else {
        console.log('✅ No broken links found!');
    }

    // Save detailed report
    const reportPath = path.join(baseDir, 'audit-report-links.json');
    fs.writeFileSync(reportPath, JSON.stringify(results, null, 2));
    console.log(`\nDetailed report saved to: ${reportPath}`);
}

// Run the audit
auditLinks();